File: Graph/P210CourseScheduleII/CourseScheduleII.java
package Graph.P210CourseScheduleII;

import java.util.*;

/**
 * @CLassName CourseScheduleII
 * @Description TODO
 * @Author cst
 * @Date 2/1/21  2:26 PM
 */
public class CourseScheduleII {
    // topological sort
    public int[] findOrder(int numCourses, int[][] prerequisites) {
        List<Integer> list = new ArrayList<>();
        Map<Integer, Set<Integer>> graph = new HashMap<>();
        int[] inDegree = new int[numCourses];
        for (int i = 0; i<numCourses; i++) {
            graph.put(i, new HashSet<>());
        }

        for (int[] prerequisite : prerequisites) {
            graph.get(prerequisite[1]).add(prerequisite[0]);
            inDegree[prerequisite[0]]++;
        }

        Queue<Integer> queue = new LinkedList<>();
        Set<Integer> visited = new HashSet<>();
        for (int i = 0; i<inDegree.length; i++) {
            if (inDegree[i] == 0) {
                queue.offer(i);
            }
        }

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i<size; i++) {
                int curr = queue.poll();
                if (visited.add(curr)) {
                    list.add(curr);
                    for (int next : graph.get(curr)) {
                        inDegree[next] -= 1;
                        if (inDegree[next] == 0) {
                            queue.offer(next);
                        }
                    }
                }
            }
        }
        int[] res = new int[list.size()];
        for (int i =0; i<list.size(); i++) {
            res[i] = list.get(i);
        }
        return res.length == numCourses ? res : new int[]{};
    }
}

File: Graph/P210CourseScheduleII/CourseScheduleII.py
import unittest
from typing import List
from collections import deque

# using topological sort to check if there ia a circle in the graph
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        indegree = [0] * numCourses
        graph, visited, count = {}, set(), 0
        res = []    
        for i in range(numCourses):
            graph[i] = []
        for pre in prerequisites:
            graph[pre[1]].append(pre[0])
            indegree[pre[0]] += 1
        
        queue = deque()
        for i in range(numCourses):
            if indegree[i] == 0:
                queue.append(i)
        while queue:
            curr = queue.popleft()
            if curr in visited:
                continue
            visited.add(curr)
            res.append(curr)
            count += 1
            for next in graph[curr]:
                indegree[next] -= 1
                if indegree[next] == 0:
                    queue.append(next)
        return res if numCourses == count else []
              
        

class UnitTest(unittest.TestCase):
    def test(self):
        numCourses = 2
        prerequisites = [[1,0]]
        self.assertEqual([0,1], Solution().findOrder(numCourses, prerequisites))
        
        numCourses = 4
        prerequisites = [[1,0],[2,0],[3,1],[3,2]]
        self.assertEqual([0,1,2,3], Solution().findOrder(numCourses, prerequisites))
        
        numCourses = 1
        prerequisites = []
        self.assertEqual([0], Solution().findOrder(numCourses, prerequisites))
if __name__ == "__main__":
    unittest.main()